<!doctype html>
<html>
<head>
    <meta charset='utf-8'>
    <title>Promise test</title>
</head>
<body>
    <h1>Promise test</h1>
    <script type="text/javascript">
    function asyncFn(callback){
      console.log("start asyncFn");
      setTimeout(function(){
        callback.call(this);
      }, 1000);
      console.log("end asyncFn");
    }

    var p1 = new Promise(function(resolve, rejected){
      resolve(1);
    }).then(function(r){  // then1
      console.log("then1.success");
      // 如果返回的是 Promise 实例，会等待此 Promise resolved 后 then2 才会执行
      return new Promise(function(resolve, rejected){
        asyncFn(function(i){
          resolve(2);
          console.log("call callback"); // 实测 Chrome 中这行语句在 then2 前输出
        }, 2);
      });
    }).then(function(r){  // then2
      console.log("then2.success");

      // then 中抛出的异常自动转化为下个 then 中的 rejected 函数的参数
      throw new Error("throw error in then2.success");
    }).then(function(){console.log("then3.success")}, function(e){console.log("then3.error - ", e)});
    </script>
</body>
</html>